### PR TITLE
polys: add RR[x].is_Exact == False

### LINKED ISSUE #25182: Question about sympy.solve / sympy.nonlinsolve
I am trying to solve the nonlinear system of equations below. (These are the three 'law of cosine' equations, I found them in the internet.)
```
a1, b1, c1, ca, cb, cg = sm.symbols('a1, b1, c1, ca, cb, cg')
eq1 = a1*a1 + b1*b1 - 2.*a1*b1*cg - c1*c1
eq2 = a1*a1 + c1*c1 - 2.*a1*c1*cb - b1*b1
eq3 = b1*b1 + c1*c1 - 2.*b1*c1*ca - a1*a1

loesung = sm.nonlinsolve([eq1, eq2, eq3], [c1, cb, cg])
```
I first tried sympy.solve.
It gave a solution, actually two solutions, but in the documentation I could not find, that sympy.solve will solve systems of nonlinear equations.
Also, the solutions seemed to be wrong, e.g. the cos(angle) > 1

Then I tried sympy.nonlinsolve.
After about 45 min of running, I aborted it.

My question:
Is it normal, that it takes so long to find a solution, or is this an indication, that none exists?

Many thanks for any explanations!

### PR DESCRIPTION
**Pull request by gschintgen**
#### References to other Issues or PRs

Fixes #25182.

#### Brief description of what is fixed or changed
This fix has been suggested in #25182. Unfortunately it introduces other test failures. I'll leave this as a draft for future reference in case anybody else wants to get this merged. (I'm not familiar with the polys code base.)

#### Release Notes

* polys
    * The domain is_Exact and get_exact methods now correctly handle composite domains like polynomial rings and `CC.get_exact()` now gives `QQ_I`.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
The test failure ultimately comes down to this change:

On master we have:
```python
In [1]: Poly(1, x).cancel(Poly(51.05*y*x - 1.0, x))
Out[1]: (1.0, Poly(1.0, x, domain='RR[y]'), Poly(51.05*y*x - 1.0, x, domain='RR[y]'))
```
Instead the PR gives:
```python
In [1]: Poly(1, x).cancel(Poly(51.05*y*x - 1.0, x))
Out[1]: 
(1.0, Poly(17592186044421.0, x, domain='RR[y]'), Poly(898081097567692.0*y*x - 17592186044421.0, x,  domain='RR[y]'))
```
This is because this check now correctly observes that `RR[y]` is not exact leading to a different codepath:
https://github.com/sympy/sympy/blob/ffd3f417742ecb9ee0cc0820677980ce78d8ea7d/sympy/polys/euclidtools.py#L1492

The end result is that `dup_rr_prs_gcd` is used instead of `dup_qq_heu_gcd` and apparently they handle swapping denominators between the two polynomials differently. This is problematic because an inexact float like 51.05 can end up having a large denominator.

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -4248,6 +4248,53 @@ def equal_valued(x, y):
         return (1 << neg_exp) == q
 
 
+def all_close(expr1, expr2, rtol=1e-5, atol=1e-8):
+    """Return True if expr1 and expr2 are numerically close.
+
+    The expressions must have the same structure, but any Rational, Integer, or
+    Float numbers they contain are compared approximately using rtol and atol.
+    Any other parts of expressions are compared exactly.
```
Comment (oscarbenjamin) [inline sympy/core/numbers.py:4256]: A function like this has been needed for a long time. Maybe a better name can be found for it though.

**Comment by oscarbenjamin** [comment]
This is ready for review.

**Comment by gschintgen** [comment]
While I'm not that knowledgeable about the `polys` module, I did look over all the commits (and in particular the added tests) and it all looks good to me.

File: sympy/polys/euclidtools.py [hunk 2]
```diff
@@ -1489,7 +1489,19 @@ def dup_inner_gcd(f, g, K):
...
 
     """
-    if not K.is_Exact:
+    # XXX: This used to check for K.is_Exact but leads to awkward results when
+    # the domain is something like RR[z] e.g.:
+    #
+    # >>> g, p, q = Poly(1, x).cancel(Poly(51.05*x*y - 1.0, x))
+    # >>> g
+    # 1.0
+    # >>> q
```
Comment (jksuom) [inline sympy/polys/euclidtools.py:1498]: Should this be p?

Comment (oscarbenjamin) [reply sympy/polys/euclidtools.py]: Yes. I've just changed it.

**Comment by jksuom** [comment]
This looks good to me.

**Comment by oscarbenjamin** [comment]
Thanks.

Does anyone have any opinions about how gcd/cancel should work for floats?

I have tried to preserve existing behaviour here but I wonder if it should be changed (in a separate PR).

Current behaviour seems inconsistent which is problematic because it means we get very different results depending on the domain:
```python
In [39]: p1 = x

In [40]: p2 = 51.05*x - 1

In [41]: cancel(p1/p2)
Out[41]: 
    1.0⋅x    
─────────────
51.05⋅x - 1.0

In [42]: Poly(p1).cancel(Poly(p2))
Out[42]: (1.0, Poly(1.0*x, x, domain='RR'), Poly(51.05*x - 1.0, x, domain='RR'))

In [43]: Poly(p1).cancel(Poly(p2, domain=QQ))
Out[43]: (17592186044421, Poly(x, x, domain='QQ'), Poly(898081097567692*x - 17592186044421, x, domain='QQ'))

In [45]: Poly(p1).cancel(Poly(p2, domain=QQ[y]))
Out[45]: (1, Poly(17592186044421*x, x, domain='QQ[y]'), Poly(898081097567692*x - 17592186044421, x, domain='QQ[y]'))

In [46]: Poly(p1).cancel(Poly(p2, domain=RR))
Out[46]: (1.0, Poly(1.0*x, x, domain='RR'), Poly(51.05*x - 1.0, x, domain='RR'))

In [47]: Poly(p1).cancel(Poly(p2, domain=RR[y]))
Out[47]: (1.0, Poly(1.0*x, x, domain='RR[y]'), Poly(51.05*x - 1.0, x, domain='RR[y]'))

In [64]: R = QQ[x]; R(p1).cancel(R(p2))
Out[64]: (17592186044421*x, 898081097567692*x - 17592186044421)

In [65]: R = RR[x]; R(p1).cancel(R(p2))
Out[65]: (x, 51.05*x - 1.0)

In [71]: p2 = Rational('51.05')*x - 1

In [72]: R = QQ[x]; R(p1).cancel(R(p2))
Out[72]: (20*x, 1021*x - 20)
```
This seems to be somewhat compounded by the fact that `QQ` fails to recognise 51.05 as an exact rational in the way that it does for some other numbers:
```python
In [68]: QQ.convert_from(RR(1/3.), RR)
Out[68]: mpq(1,3)

In [70]: QQ.convert_from(RR(0.1), RR)
Out[70]: mpq(1,10)

In [69]: QQ.convert_from(RR(51.05), RR)
Out[69]: mpq(898081097567692,17592186044421)

In [83]: RR(1+1e-15)
Out[83]: 1.0

In [84]: RR(1+1e-14)
Out[84]: 1.00000000000001

In [85]: RR(1+1e-13)
Out[85]: 1.0000000000001

In [86]: cancel((x**2 - 1 + 1e-15)/(x - 1.0))
Out[86]: 1.0⋅x + 1.0

In [87]: cancel((x**2 - 1 + 1e-14)/(x - 1.0))
Out[87]: 1.0⋅x + 1.0

In [88]: cancel((x**2 - 1 + 1e-13)/(x - 1.0))
Out[88]: 
     2                  
1.0⋅x  - 0.9999999999999
────────────────────────
      1.0⋅x - 1.0
```
On a basic level it seems that gcd does not work the same way for RR:
```python
In [91]: primitive(2*x**2 - 2.0)
Out[91]: 
⎛          2      ⎞
⎝1.0, 2.0⋅x  - 2.0⎠

In [92]: primitive(2*x**2 - 2)
Out[92]: 
⎛    2    ⎞
⎝2, x  - 1⎠

In [93]: RR.gcd(RR(1/4), RR(1/2))
Out[93]: 1.0

In [94]: QQ.gcd(QQ(1/4), QQ(1/2))
Out[94]: mpq(1,4)
```

I wonder what is a good way to define what `cancel` should do for RR. Generally the purpose of `cancel` is to place the numerator and denominator into a canonical form. One part of that is extracting the gcd from the numerator and denominator. Another part is making them canonical with respect to multiplication by units.

For QQ we can choose units to make the numerator and denominator primitive with integer coefficients. In the case of RR doing the same means extracting a large power of 2 denominator which is probably not ideal. An alternative could be to make the denominator have `1` as its leading coefficient. That might be bad though if the leading coefficient is something like `1e-16`.

Maybe the best approach here is to try somehow to avoid any multiplication by units. Then we don't get a canonical fraction but we can at least preserve the values of the floats as much as possible.

**Pull request merged**